src/app/features/data-import/data-import.service.ts

```diff
@@ -20,69 +32,154 @@ export class DataImportService {
+    if (!confirmed) {
+      return;
+    }
 
-    const dialogRef = this.confirmationDialog.openDialog(
-      $localize`Import new data?`,
-      $localize`Are you sure you want to import this file? This will add or update ${
-        newData.trim().split("\n").length - 1
-      } records from the loaded file. Existing records with same "_id" in the database will be overwritten!`
-    );
+    if (importMeta.transactionId) {
+      await this.deleteExistingRecords(importMeta);
+    }
+
+    await this.importCsvContentToDB(csvFile, importMeta);
 
-    dialogRef.afterClosed().subscribe(async (confirmed) => {
-      if (!confirmed) {
-        return;
+    const snackBarRef = this.snackBar.open(
+      $localize`Import completed?`,
+      $localize`Undo`,
+      {
+        duration: 8000,
       }
+    );
+    snackBarRef
+      .onAction()
+      .pipe(untilDestroyed(this))
+      .subscribe(async () => {
+        await this.backupService.clearDatabase();
+        await this.backupService.importJson(restorePoint, true);
+      });
+  }
 
-      await this.importCsvContentToDB(newData);
+  private getUserConfirmation(
+    csvFile: ParseResult,
+    importMeta: ImportMetaData
+  ): Promise<boolean> {
+    const refTitle = $localize`Import new data?`;
+    let refText = $localize`Are you sure you want to import this file?
+      This will add or update ${csvFile.data.length} records from the loaded file.`;
+    if (importMeta.transactionId) {
+      refText = $localize`${refText} All existing records imported with the transaction id '${importMeta.transactionId}' will be deleted!`;
+    }
+    const dialogRef = this.confirmationDialog.openDialog(refTitle, refText);
 
-      const snackBarRef = this.snackBar.open(
-        $localize`Import completed?`,
-        "Undo",
-        {
-          duration: 8000,
-        }
-      );
-      snackBarRef
-        .onAction()
-        .pipe(untilDestroyed(this))
-        .subscribe(async () => {
-          await this.backupService.clearDatabase();
-          await this.backupService.importJson(restorePoint, true);
-        });
+    return new Promise<boolean>((resolve) => {
+      dialogRef.afterClosed().subscribe((confirmed) => resolve(confirmed));
     });
   }
+
+  private async deleteExistingRecords(importMeta: ImportMetaData) {
+    const existing = await this.db.getAll(
+      Entity.createPrefixedId(importMeta.entityType, importMeta.transactionId)
+    );
+    return Promise.all(existing.map((entity) => this.db.remove(entity)));
+  }
+
+  private async importCsvContentToDB(
+    csv: ParseResult,
+    importMeta: ImportMetaData
+  ): Promise<void> {
+    for (const row of csv.data) {
+      const entity = this.createEntityWithRowData(row, importMeta);
+      this.createSearchIndices(importMeta, entity);
+      if (!entity["_id"]) {
+        entity["_id"] = `${importMeta.entityType}:${
+          importMeta.transactionId
+        }-${uuid().substr(9)}`;
+      }
+      await this.db.put(entity, true);
```

Comment:
Couldn't we instantiate a new entity and save that with EntityMapper instead of raw db.put? That would also take care of things like searchIndices out of the box

Reply:
Lets do this together with more a more advanced column mapping when we bring the property schemas into action. I note it in #1135